src/inmanta/ast/statements/assign.py

```diff
@@ -581,8 +601,45 @@ def execute(self, requires: typing.Dict[object, object], resolver: Resolver, que
 
         return result_string
 
-    def get_dataflow_node(self, graph: DataflowGraph) -> dataflow.NodeReference:
-        return dataflow.NodeStub("StringFormat.get_node() placeholder for %s" % self).reference()
 
-    def __repr__(self) -> str:
-        return "Format(%s)" % self._format_string
+class FStringFormatter(Formatter):
+    def __init__(self) -> None:
+        Formatter.__init__(self)
+
+    def get_field(self, key: str, args: Sequence[object], kwds: Mapping[str, object]) -> Tuple[object, str]:
+        """
+        Overrides Formatter.get_field. Composite variable names are expected to be resolved at this point and can be
+        retrieved by their full name.
+        """
+        return (kwds[key], key)
+
+
+class StringFormatV2(FormattedString):
+    """
+    Create a new string by using python build in formatting
+    """
+
+    __slots__ = ()
+
+    def __init__(self, format_string: str, variables: Sequence[typing.Tuple["Reference", str]]) -> None:
+        only_refs: Sequence["Reference"] = [k for (k, _) in variables]
+        super().__init__(format_string, only_refs)
+        self._variables = only_refs
+
+    def execute(self, requires: typing.Dict[object, object], resolver: Resolver, queue: QueueScheduler) -> object:
+        super().execute(requires, resolver, queue)
+        formatter: FStringFormatter = FStringFormatter()
```

Comment:
Sadly the underlying formatter has no ability to cache the parsed string, so we have to re-parse every time. We might be able to squeeze a bit more performance out of it, but alas. 

Perhaps you could add a comment here to clarify why we don't cache the `formatter`

Reply:
I went with `        # We can't cache the formatter because it has no ability to cache the parsed string`, does this make sense ? I must admit it I don't fully get your point still.